java-checks/src/main/java/org/sonar/java/checks/ObjectCreatedOnlyToCallGetClassCheck.java

```diff
@@ -0,0 +1,119 @@
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.checks.methods.MethodInvocationMatcher;
+import org.sonar.java.checks.methods.TypeCriteria;
+import org.sonar.java.model.AbstractTypedTree;
+import org.sonar.java.resolve.Symbol;
+import org.sonar.java.resolve.Symbol.TypeSymbol;
+import org.sonar.java.resolve.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S2133",
+  name = "Objects should not be created only to \"getClass\"",
+  tags = {"performance"},
+  priority = Priority.BLOCKER)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.FAULT_TOLERANCE)
+@SqaleConstantRemediation("2min")
+public class ObjectCreatedOnlyToCallGetClassCheck extends AbstractMethodDetection {
+
+  @Override
+  protected List<MethodInvocationMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodInvocationMatcher.create().typeDefinition(TypeCriteria.subtypeOf("java.lang.Object")).name("getClass"));
+  }
+
+  @Override
+  protected void onMethodFound(MethodInvocationTree mit) {
+    if (hasSemantic() && mit.methodSelect().is(Kind.MEMBER_SELECT)) {
+      ExpressionTree expressionTree = ((MemberSelectExpressionTree) mit.methodSelect()).expression();
+      if (expressionTree.is(Kind.NEW_CLASS)) {
+        reportIssue(expressionTree);
+      } else if (expressionTree.is(Kind.IDENTIFIER) && variableUsedOnlyToGetClass((IdentifierTree) expressionTree)) {
+        reportIssue(getInitializer((IdentifierTree) expressionTree));
+      }
+    }
+  }
+
+  private ExpressionTree getInitializer(IdentifierTree tree) {
+    Symbol symbol = getSemanticModel().getReference(tree);
+    return ((VariableTree) getSemanticModel().getTree(symbol)).initializer();
+  }
+
+  private boolean variableUsedOnlyToGetClass(IdentifierTree tree) {
+    if ("this".equals(tree.name()) || "super".equals(tree.name())) {
+      return false;
+    }
+    Symbol symbol = getSemanticModel().getReference(tree);
+    return getSemanticModel().getUsages(symbol).size() == 1 && hasBeenInitialized(symbol);
+  }
+
+  private boolean hasBeenInitialized(Symbol symbol) {
+    VariableTree source = (VariableTree) getSemanticModel().getTree(symbol);
+    return source.initializer() != null;
```

Comment:
As discussed, this should test that initializer is NewClassTree otherwise it can get the reference to the object from elsewhere.